### PR TITLE
add pyspark support to tree surrogate model in error analysis

### PR DESCRIPTION
**Pull request by imatiach-msft**
## Description

Add pyspark support to tree surrogate model in error analysis.  This is part 1 of a set of PRs to add pyspark support to ErrorAnalysis backend package and the visualization dashboard.

Some things not yet supported in this PR:
- handling all metrics currently available in error analysis (this PR only supports Error Rate)
- handling categoricals

## Areas changed

npm packages changed:

- [ ] responsibleai/causality
- [ ] responsibleai/core-ui
- [ ] responsibleai/counterfactuals
- [ ] responsibleai/dataset-explorer
- [ ] responsibleai/fairness
- [ ] responsibleai/interpret
- [ ] responsibleai/localization
- [ ] responsibleai/mlchartlib
- [ ] responsibleai/model-assessment

Python packages changed:

- [ ] raiwidgets
- [ ] responsibleai
- [x] erroranalysis
- [ ] rai_core_flask

## Tests

- [ ] No new tests required.
- [ ] New tests for the added feature are part of this PR.
- [ ] I validated the changes manually.

## Screenshots (if appropriate):
Fully working pyspark example, this is the first step towards this POC:

image

## Documentation:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by vinuthakaranth** [comment]
Just curious will this be added later in point of time?

- handling all metrics currently available in error analysis (this PR only supports Error Rate)
- handling categoricals

**Comment by vinuthakaranth** [comment]
Just curious will this be added later in point of time?

- handling all metrics currently available in error analysis (this PR only supports Error Rate)
- handling categoricals

**Comment by imatiach-msft** [comment]
"Just curious will this be added later in point of time?"
Yes, these will be added at a later point in time.  Certainly handling categoricals will be added and should be straightforward, just a lot more code.  Adding more metrics may be a bit more complicated, as we will need to find all corresponding metrics calculations in spark.

File: erroranalysis/erroranalysis/_internal/cohort_filter.py [hunk 1]
```diff
@@ -32,12 +33,13 @@ def filter_from_cohort(analyzer, filters, composite_filters):
...
-        df = pd.DataFrame(df, columns=feature_names)
-    else:
-        # Note: we make a non-deep copy of the input DataFrame since
-        # we will add columns below
-        df = df.copy()
+    if not is_spark(df):
+        if not isinstance(df, pd.DataFrame):
+            df = pd.DataFrame(df, columns=feature_names)
+        else:
+            # Note: we make a non-deep copy of the input DataFrame since
```
Comment (romanlutz) [inline erroranalysis/erroranalysis/_internal/cohort_filter.py:40]: a shallow copy? Not sure what non-deep is or if it's the same.

File: erroranalysis/erroranalysis/_internal/surrogate_error_tree.py [hunk 2]
```diff
@@ -182,20 +254,97 @@ def compute_error_tree(analyzer,
...
+        DIFF,
+        F.when(F.col(analyzer.true_y) != F.col(PREDICTION),
+               1).otherwise(0))
+    if analyzer.model_task == ModelTask.CLASSIFICATION:
+        diff_data = diff_data.drop(PREDICTION, RAW_PREDICTION, PROBABILITY)
+    else:
+        diff_data = diff_data.drop(PREDICTION)
+    model = create_surrogate_model_pyspark(analyzer, diff_data, max_depth,
+                                           num_leaves, min_child_samples)
+    # TODO: update lightgbm in pyspark to get around file requirement
```
Comment (romanlutz) [inline erroranalysis/erroranalysis/_internal/surrogate_error_tree.py:296]: Is that a separate PR somewhere?

**Review by romanlutz — APPROVED** [review]

File: erroranalysis/erroranalysis/_internal/surrogate_error_tree.py (hunk 2 above)
Comment (imatiach-msft) [reply erroranalysis/erroranalysis/_internal/surrogate_error_tree.py:296]: yes I will send a PR in https://github.com/microsoft/SynapseML/ for this at some point

File: erroranalysis/erroranalysis/_internal/cohort_filter.py (hunk 1 above)
Comment (imatiach-msft) [reply erroranalysis/erroranalysis/_internal/cohort_filter.py:40]: oh, yes, that's what you call it I guess.  I will update comment in next PR.

**Pull request merged**